syntax = "proto3";

option go_package = "github.com/carnei-ro/my-crud-grpc/protos;protos";
option java_multiple_files = true;
option java_package = "ro.carnei.mycrudgrpc.protos";
option java_outer_classname = "MyCrudGrpcProtos";
option objc_class_prefix = "MYCRUD";

// import "google/api/annotations.proto";

package mycrudgrpc;

// The service definition.

service Movie {
  rpc GetMovies (Empty) returns (stream MovieInfo) {
    option (google.api.http) = {
      get: "/v1/movies"
      body: "*"
    };
  }
  rpc GetMovie (Uuid) returns (MovieInfo) {
    option (google.api.http) = {
      get: "/v1/movies/{uuid}"
      body: "*"
    };
  }
  rpc CreateMovie (MovieInfo) returns (Uuid) {
    option (google.api.http) = {
      post: "/v1/movies"
      body: "*"
    };
  }
  rpc UpdateMovie (MovieInfo) returns (Status) {
    option (google.api.http) = {
      put: "/v1/movies/{uuid}"
      body: "*"
    };
  }
  rpc DeleteMovie (Uuid) returns (Status) {
    option (google.api.http) = {
      delete: "/v1/movies/{uuid}"
      body: "*"
    };
  }
}

message Empty {}

message MovieInfo {
  string uuid = 1;
  string title = 2;
  string description = 3;
  Director director = 4;
}

message Director {
  string name = 1;
  string surname = 2;
}

message Uuid {
  string uuid = 1;
}

message Status {
  bool success = 1;
}
